test/model/src/include/model/core.h

```diff
@@ -0,0 +1,437 @@
+     *     Compare to another data value.
+     */
+    inline bool
+    operator!=(const data_value &other) const noexcept
+    {
+        return !(*this == other);
+    }
+
+    /*
+     * data_value::operator< --
+     *     Compare to another data value.
+     */
+    inline bool
+    operator<(const data_value &other) const noexcept
+    {
+        return _data < other._data;
+    }
+
+    /*
+     * data_value::operator<= --
+     *     Compare to another data value.
+     */
+    inline bool
+    operator<=(const data_value &other) const noexcept
+    {
+        return _data <= other._data;
+    }
+
+    /*
+     * data_value::operator> --
+     *     Compare to another data value.
+     */
+    inline bool
+    operator>(const data_value &other) const noexcept
+    {
+        return !(*this <= other);
+    }
+
+    /*
+     * data_value::operator> --
+     *     Compare to another data value.
+     */
+    inline bool
+    operator>=(const data_value &other) const noexcept
+    {
+        return !(*this < other);
+    }
+
+    /*
+     * data_value::tombstone --
+     *     Check if this is a None value.
+     */
+    inline bool
+    none() const noexcept
+    {
+        return !_data.has_value();
+    }
+
+private:
+    /*
+     * data_value::data_value --
+     *     Create a new instance.
+     */
+    inline data_value() : _data(std::nullopt) {}
+
+    std::optional<std::string> _data;
+};
+
+/*
+ * NONE --
+ *     The "None" value.
+ */
+extern const data_value NONE;
+
+/*
+ * kv_update --
+ *     The data value stored in a KV table, together with the relevant update information, such as
+ *     the timestamp.
+ */
+class kv_update {
+
+public:
+    /*
+     * kv_update::timestamp_comparator --
+     *     The comparator that uses timestamps only.
+     */
+    struct timestamp_comparator {
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare two updates.
+         */
+        bool
+        operator()(const kv_update &left, const kv_update &right) const noexcept
+        {
+            return left._timestamp < right._timestamp;
+        }
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare the update to the given timestamp.
+         */
+        bool
+        operator()(const kv_update &left, timestamp_t timestamp) const noexcept
+        {
+            return left._timestamp < timestamp;
+        }
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare the update to the given timestamp.
+         */
+        bool
+        operator()(timestamp_t timestamp, const kv_update &right) const noexcept
+        {
+            return timestamp < right._timestamp;
+        }
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare two updates.
+         */
+        bool
+        operator()(const kv_update *left, const kv_update *right) const noexcept
```

Comment:
I think this is problematic: as `nullptr` is a valid value for `const kv_update *`.

Reply:
Agree. For now I just added nullptr handling to the comparators. I ended up using pointers to `kv_update` in the `deque`, so that the caller can safely keep a pointer (or a reference) to an instance of `kv_update`, while things are being added to the `deque`. Consequently, I ended up having to deal with pointers.